src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.LazySyntaxTree.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    public partial class CSharpSyntaxTree
+    {
+        private sealed class LazySyntaxTree : CSharpSyntaxTree
+        {
+            private readonly SourceText _text;
+            private readonly CSharpParseOptions _options;
+            private readonly string _path;
+            private readonly ImmutableDictionary<string, ReportDiagnostic> _diagnosticOptions;
+            private CSharpSyntaxNode? _lazyRoot;
+
+            internal LazySyntaxTree(
+                SourceText text,
+                CSharpParseOptions options,
+                string path,
+                ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions)
+            {
+                Debug.Assert(options != null);
+
+                _text = text;
+                _options = options;
+                _path = path ?? string.Empty;
+                _diagnosticOptions = diagnosticOptions ?? EmptyDiagnosticOptions;
+            }
+
+            public override string FilePath
+            {
+                get { return _path; }
+            }
+
+            public override SourceText GetText(CancellationToken cancellationToken)
+            {
+                return _text;
+            }
+
+            public override bool TryGetText([NotNullWhen(true)] out SourceText? text)
+            {
+                text = _text;
+                return true;
+            }
+
+            public override Encoding? Encoding
+            {
+                get { return _text.Encoding; }
+            }
+
+            public override int Length
+            {
+                get { return _text.Length; }
+            }
+
+            public override CSharpSyntaxNode GetRoot(CancellationToken cancellationToken)
+            {
+                if (!TryGetRoot(out var root))
+                {
+                    // Parse the syntax tree
+                    var tree = SyntaxFactory.ParseSyntaxTree(_text, _options, _path, cancellationToken);
+                    root = CloneNodeAsRoot((CSharpSyntaxNode)tree.GetRoot(cancellationToken));
+
+                    // Lazily initialize _lazyRoot, and use the first instance successfully written
+                    root = Interlocked.CompareExchange(ref _lazyRoot, root, null) ?? root;
+                }
+
+                return root;
+            }
+
+            public override bool TryGetRoot([NotNullWhen(true)] out CSharpSyntaxNode? root)
```

Comment:
>TryGetRoot [](start = 33, length = 10)

Are we testing that the `SyntaxTree` from `GeneratorDriver` is parsed lazily? Perhaps something like:
```
var driver = new CSharpGeneratorDriver(...);
var tree = driver.ParseGeneratedSourceTest(...);

Assert.False(tree.TryGetRoot(out var root));
Assert.Null(root);

_ = tree.GetRoot(default);
Assert.True(tree.TryGetRoot(out root));
Assert.NotNull(root);
```

Reply:
Tests added.